### PR TITLE
Implementation of the linear model

### PR DESCRIPTION
**Pull request by oleksiyskononenko**
This is an initial implementation of the linear model with stochastic gradient descent learning. It supports binomial and multinomial regressions, as well as regression for continuous targets. API is similar to the one for the existing FTRL model. 

Still needs to be done:
- add documentation;
- resolve issues with several tests, that use training datasets with string columns;
- better refactoring to avoid code duplication between FTRL and the linear model.

WIP for #2358

### CODE REVIEW

File: src/core/models/dt_linearmodel.cc [hunk 1]
```diff
@@ -0,0 +1,1096 @@
...
+
+        // Training.
+        dt::nested_for_static(iteration_size, ChunkSize(MIN_ROWS_PER_THREAD), [&](size_t i) {
+          size_t ii = (iteration_start + i) % dt_X_train->nrows();
+          U target;
+          bool isvalid = target_col0_train.get_element(ii, &target);
+
+          if (isvalid && _isfinite(target) && read_row(ii, cols, x)) {
+            // Loop over all the labels
+            for (size_t k = 0; k < label_ids_train.size(); ++k) {
```
Comment (st-pasha) [inline src/core/models/dt_linearmodel.cc:432]: I don't think that's how multinomial regression is supposed to work... 
I'm more used to log-likelihood formulation, but basically in a multinomial regression the particular value of `y_i` in the current row determines which term will be added to the total likelihood that will in the end be maximized. So it's 
`log-likelihood = Sum[ log Pr(x_i | y_i) ]`. 
I mean, there shouldn't be a for-loop over all labels -- just one label, `y_i` is in play here.

File: src/core/models/dt_linearmodel.cc [hunk 2]
```diff
@@ -0,0 +1,1096 @@
...
+                           [&](size_t f_id, T f_imp) {
+                             fi[f_id] += f_imp;
+                            }
+                         ));
+                T y = static_cast<T>(targetfn(target, label_ids_train[k]));
+                grad = 2 * dlinkfn(grad) * (grad - y);
+                if (j) grad *= x[j - 1];
+                grad += copysign(lambda1, w[k][j]); // L1 regularization
+                grad += 2 * lambda2 * w[k][j];      // L2 regularization
+                w[k][j] -= eta * grad;
```
Comment (st-pasha) [inline src/core/models/dt_linearmodel.cc:446]: Are you updating the model's coefficients on the go? This is definitely not the standard way of computing regressions, so a link to a paper would be necessary. I saw https://www.jmlr.org/papers/volume15/dhurandhar14a/dhurandhar14a.pdf, but that's probably not what you were implementing.

File: src/core/models/column_caster.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      case dt::SType::BOOL:
+      case dt::SType::INT8:    col_casted = make_casted_column<int8_t, T>(col, stype); break;
+      case dt::SType::INT16:   col_casted = make_casted_column<int16_t, T>(col, stype); break;
+      case dt::SType::INT32:   col_casted = make_casted_column<int32_t, T>(col, stype); break;
+      case dt::SType::INT64:   col_casted = make_casted_column<int64_t, T>(col, stype); break;
+      case dt::SType::FLOAT32: col_casted = make_casted_column<float, T>(col, stype); break;
+      case dt::SType::FLOAT64: col_casted = make_casted_column<double, T>(col, stype); break;
+      default: throw TypeError() << "Columns with stype `" << stype << "` are not supported";
+    }
+    cols_casted.push_back(std::move(col_casted));
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/column_caster.h:71]: `make_casted_column()` also checks for finiteness and produces NA for infinite values, will `col.cast()` do that? I'm not sure.

File: src/core/models/column_caster.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+           [](T_from x, bool x_isvalid, T_to* out) {
+             *out = static_cast<T_to>(x);
+             return x_isvalid && _isfinite(x);
+           },
+           col.nrows(),
+           stype
+         ));
+
+  return col_casted;
+}
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/column_caster.h:45]: How could I use it to convert infinity to NA?

File: src/core/models/dt_linearmodel.cc [hunk 5]
```diff
@@ -0,0 +1,1096 @@
...
+#include "models/column_caster.h"
+#include "models/dt_linearmodel.h"
+#include "parallel/api.h"
+#include "parallel/atomic.h"
+#include "progress/work.h"      // dt::progress::work
+#include "wstringcol.h"
+#include "utils/macros.h"
+#if DT_OS_WINDOWS
+  #undef copysign
+#endif
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.cc:35]: May be, but may be not. I remember dealing with copysign was a total mess on Windows, because it was implicitly defined in very unexpected places. I can try to move it to `macros.h` to see what happens.

File: src/core/models/dt_linearmodel_base.h [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+namespace dt {
+
+
+/**
+ *  Supported linear model types.
+ */
+enum class LinearModelType : size_t {
+  NONE        = 0, // Untrained model
+  AUTO        = 1, // Automatically detect model type
+  REGRESSION  = 2, // Numerical regression
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel_base.h:35]: It is a linear regression, but not Ordinary Least Squares. It is a linear regression with SGD (Stochastic Gradient Descent) learning.

File: src/core/models/dt_linearmodel.h [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+namespace dt {
+
+
+/**
+ *  Class template that implements all the virtual methods declared in
+ *  dt::LinearModelBase.
+ */
+template <typename T /* float or double */>
+class LinearModel : public dt::LinearModelBase {
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.h:38]: Yes, not needed. Left from some old times as well as for FTRL. Will remove it.

I had to create a base class here due to `py::LinearModel`. This class is not templated, so keeps a pointer to `dt::LinearModelBase`. Cannot use a pointer to `LinearModel<T>`, because there is no `T` defined.

File: src/core/models/dt_linearmodel_base.h [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+#include "options.h"
+#include "str/py_str.h"
+namespace dt {
+
+
+/**
+ *  Supported linear model types.
+ */
+enum class LinearModelType : size_t {
+  NONE        = 0, // Untrained model
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel_base.h:33]: Yes, it is a separate flag, I just put it here for convenience. Indeed any of the model types can be in any state:

```
>>> lm.model_type
'auto'
>>> lm.model_type_trained
'none'
```

File: src/core/models/dt_linearmodel.h [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    size_t : 56;
+    // Model's shape is (nfeatures + 1, nlabels)
+    dtptr dt_model;
+    std::vector<T*> w;
+    LinearModelType model_type;
+
+    // Feature importances datatable of shape (nfeatures, 2),
+    // where the first column contains feature names and the second one
+    // feature importance values.
+    dtptr dt_fi;
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.h:51]: Yeah, it is also the same for FTRL, I had to invent some definition. Probably also need to update documentation.

File: src/core/models/dt_linearmodel.h [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    LinearModelParams params;
+
+    // Individual parameters converted to T type.
+    T eta;
+    T lambda1;
+    T lambda2;
+    T nepochs;
+
+    // Vector of feature interactions.
+    std::vector<sztvec> interactions;
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.h:63]: Yes, could be like that also. I believe that when I first added this parameter to the model (FTRL at that time), we didn't have virtual columns, so having `interactions` parameter could save a lot from the memory perspective and also from the hashing perspective. In this linear model I kept it because it seems pretty convenient for the user to simply pass the interactions as a list of lists, instead of adjusting the frame.

File: src/core/models/dt_linearmodel.h [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+ */
+template <typename T /* float or double */>
+class LinearModel : public dt::LinearModelBase {
+  private:
+    // SType that corresponds to `T`
+    const SType stype;
+    size_t : 56;
+    // Model's shape is (nfeatures + 1, nlabels)
+    dtptr dt_model;
+    std::vector<T*> w;
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.h:45]: Each element of this vector corresponds to the model coefficients for one particular label. The size of each vector element is `nfeatures + 1`, because we also have an intercept term. Not sure why the size should be different in the multinomial case? Multinomial classification here ends up as a set of binomial classifiers.

File: src/core/models/dt_linearmodel.cc [hunk 12]
```diff
@@ -0,0 +1,1096 @@
...
+ *  Create labels (in the case of numeric regression there is no actual
+ *  labeles, so we just use a column name for this purpose),
+ *  set up identity mapping between models and the incoming label inficators,
+ *  call to the main `fit` method.
+ */
+template <typename T>
+template <typename U>
+LinearModelFitOutput LinearModel<T>::fit_regression() {
+  xassert(dt_y_train->ncols() == 1);
+  if (is_model_trained() && model_type != LinearModelType::REGRESSION) {
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.cc:197]: We need to support batch training I believe.

File: src/core/models/dt_linearmodel.cc [hunk 13]
```diff
@@ -0,0 +1,1096 @@
...
+ *  Fit model on a datatable.
+ */
+template <typename T>
+template <typename U, typename V> /* target column(s) data type */
+LinearModelFitOutput LinearModel<T>::fit(T(*linkfn)(T),
+                                         T(*dlinkfn)(T),
+                                         U(*targetfn)(U, size_t),
+                                         V(*targetfn_val)(V, size_t),
+                                         T(*lossfn)(T, T))
+{
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.cc:352]: May be I am using bad naming convention here. What you call `loss` is called `linkfn` and there is indeed `dlinkfn` that is a derivative. The `lossfn` is only used to estimate convergence of the model on the validation dataset.

File: src/core/models/dt_linearmodel.cc (hunk 1 above)
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.cc:432]: For the label `y_i` I train its binomial classifier on a positive, and for other classifiers I train on negative.

File: src/core/models/dt_linearmodel.cc (hunk 2 above)
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.cc:446]: Isn't it a standard SGD method? https://en.wikipedia.org/wiki/Stochastic_gradient_descent#Iterative_method

File: src/core/models/dt_linearmodel.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (st-pasha) [reply src/core/models/dt_linearmodel.h:63]: If the model doesn't need to know that these are special "interactions" columns (and it usually doesn't), then a cleaner solution would be to separate these 2 concerns: perhaps create a separate function that augments the frame with interactions, and then let the user call that function if they want to.
 After all, creating interactions could use many different settings, and it just makes no sense to put all that complexity into the model itself.

File: src/core/models/dt_linearmodel.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (st-pasha) [reply src/core/models/dt_linearmodel.h:45]: ah yes, if the intercept is implicit then one more parameter is needed. 
For multinomial, one of the coefficients is always set to 0 because otherwise they are linearly dependent: https://en.wikipedia.org/wiki/Multinomial_logistic_regression#As_a_set_of_independent_binary_regressions

File: src/core/models/dt_linearmodel.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (st-pasha) [reply src/core/models/dt_linearmodel.cc:352]: The link function is applied to `w*x_i`, I presume the loss function in your case is quadratic then? That is, are you minimizing `Sum[(y_i - link(w*x_i))^2]`?

File: src/core/models/dt_linearmodel.cc (hunk 2 above)
Comment (st-pasha) [reply src/core/models/dt_linearmodel.cc:446]: Ah yes, I didn't recognize it. Is it really better in the context of a linear regression?

File: src/core/models/dt_linearmodel.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (st-pasha) [reply src/core/models/dt_linearmodel.h]: ok, makes sense

File: src/core/models/column_caster.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (st-pasha) [reply src/core/models/column_caster.h:71]: But is it even necessary? Suppose you have infinities in some `x_it` or `y_t`. Then when computing `y_t - linkfn(w*x_t)` you might produce an infinite value (not necessary, though, because the link function may convert an infinity into something nice). At that point you can check for infinity/nan and decline to add it to the total. Note that you'd generally do that anyways, since the link function may in theory produce an infinity even for regular values of `x_t`s. Thus, there may be no benefit in preemptively replacing infinities in x's/y's.

File: src/core/models/column_caster.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (st-pasha) [reply src/core/models/column_caster.h:45]: The reason I'm strongly suggesting using the standard casting functionality is because it's already relatively hard to implement efficiently, so having it in multiple places would just make it even harder.

File: src/core/models/column_caster.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/column_caster.h:71]: TODO: Check gradient for finitneess right before updating weights.

File: src/core/models/dt_linearmodel_base.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel_base.h:35]: TODO: add links to algos and formulas in the code.

File: src/core/models/dt_linearmodel_base.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (st-pasha) [reply src/core/models/dt_linearmodel_base.h:33]: `NONE` -> `UNKNOWN` ?

[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel_base.h:33]: TODO: think if we really need to keep "auto" when the model is build

File: src/core/models/dt_linearmodel.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.h:51]: TODO: explore feature importance, rename `w` to `beta`

File: src/core/models/dt_linearmodel.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.h:63]: TODO: leave interactions creation to the user, also check FTRL

File: src/core/models/dt_linearmodel.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.h:45]: TODO: check if we want to implement multinomial logistic regression

File: src/core/models/dt_linearmodel.cc [hunk 14]
```diff
@@ -0,0 +1,1096 @@
...
+
+  xassert(model_type_train != LinearModelType::NONE);
+  switch (model_type_train) {
+    case LinearModelType::AUTO :        switch (stype_y) {
+                                        case SType::VOID:
+                                        case SType::BOOL:    res = fit_binomial(); break;
+                                        case SType::INT8:    res = fit_regression<int8_t>(); break;
+                                        case SType::INT16:   res = fit_regression<int16_t>(); break;
+                                        case SType::INT32:   res = fit_regression<int32_t>(); break;
+                                        case SType::INT64:   res = fit_regression<int64_t>(); break;
```
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.cc:109]: TODO: consider splitting this class into three

**Comment by oleksiyskononenko** [comment]
This fails
```>>> lm = LinearModel()
>>> lm.fit(dt.Frame(range(100)), dt.Frame(range(100)))
LinearModelFitOutput(epoch=1.0, loss=None)
>>> lm.model
   |      C0
   | float32
-- + -------
 0 |      NA
 1 |      NA
[2 rows x 1 column]
```

File: src/core/models/dt_linearmodel.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.cc:197]: TODO: check for efficient way of parallelizing SGD

File: src/core/models/dt_linearmodel.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/dt_linearmodel.cc:352]: TODO: check how `logloss` is related to the OLS

File: src/core/models/column_caster.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (oleksiyskononenko) [reply src/core/models/column_caster.h:45]: I guess I recall why I needed this inf check casting. The reason is that sometimes I need min/max, at least in the aggregator, and if there is an inf value it just spoils all the stats. Could be fixed if we had `min2` and `max2` functions implemented.

**Comment by ledell** [comment]
Hey! I am curious to see you're adding a linear model to `datatable`.  Is there plans to expand the this package beyond data management (into ML), or is this linear model used in service of some data task?

**Comment by oleksiyskononenko** [comment]
Hey @ledell, we already have [FTRL model](https://datatable.readthedocs.io/en/latest/api/models/ftrl.html) for ML, so just adding one more. We're not going to have a full set of ML tools, just the ones that need to be highly parallel and close to the data from DAI perspective.

**Review by st-pasha — APPROVED** [review]
